nanofi/src/sitetosite/CRawSocketProtocol.c

```diff
@@ -0,0 +1,1026 @@
+
+  int ret = writeUTF(client->_commsIdentifier, strlen(client->_commsIdentifier), False, client->_peer->_stream);
+
+  if (ret <= 0) {
+    return -1;
+  }
+
+  uint32_t prop_size;
+  PropertyValue *current, *tmp, * properties = NULL;
+
+  current = (PropertyValue *)malloc(sizeof(PropertyValue));
+
+  current->name = HandShakePropertyStr[GZIP];
+  strncpy(current->value, "false", strlen("false") +1);
+
+  HASH_ADD_KEYPTR(hh, properties, current->name, strlen(current->name), current);
+
+  current = (PropertyValue *)malloc(sizeof(PropertyValue));
+
+  current->name = HandShakePropertyStr[PORT_IDENTIFIER];
+  strncpy(current->value, client->_port_id_str, strlen(client->_port_id_str) +1);
+
+  HASH_ADD_KEYPTR(hh, properties, current->name, strlen(current->name), current);
+
+  current = (PropertyValue *)malloc(sizeof(PropertyValue));
+
+  current->name = HandShakePropertyStr[REQUEST_EXPIRATION_MILLIS];
+  sprintf(current->value, "%llu", client->_timeout);
+
+  HASH_ADD_KEYPTR(hh, properties, current->name, strlen(current->name), current);
+
+  prop_size = 3;
+
+  if (client->_currentVersion >= 5) {
+    if (client->_batch_count > 0) {
+      current = (PropertyValue *)malloc(sizeof(PropertyValue));
+
+      current->name = HandShakePropertyStr[BATCH_COUNT];
+      sprintf(current->value, "%llu", client->_batch_count);
+
+      HASH_ADD_KEYPTR(hh, properties, current->name, strlen(current->name), current);
+
+      prop_size++;
+    }
+    if (client->_batch_size > 0) {
+      current = (PropertyValue *)malloc(sizeof(PropertyValue));
+
+      current->name = HandShakePropertyStr[BATCH_SIZE];
+      sprintf(current->value, "%llu", client->_batch_size);
+
+      HASH_ADD_KEYPTR(hh, properties, current->name, strlen(current->name), current);
+
+      prop_size++;
+    }
+    if (client->_batch_duration > 0) {
+      current = (PropertyValue *)malloc(sizeof(PropertyValue));
+
+      current->name = HandShakePropertyStr[BATCH_DURATION];
+      sprintf(current->value, "%llu", client->_batch_duration);
+
+      HASH_ADD_KEYPTR(hh, properties, current->name, strlen(current->name), current);
+
+      prop_size++;
+    }
+  }
+
+  if (client->_currentVersion >= 3) {
+
+    //ret = client->_peer->writeUTF(client->_peer->getURL());
+    const char * urlstr = getURL(client->_peer);
+    ret = writeUTF(urlstr, strlen(urlstr), False, client->_peer->_stream);
+    if (ret <= 0) {
+      return -1;
+    }
+  }
+
+  ret = write_uint32_t(prop_size, client->_peer->_stream);
+  if (ret <= 0) {
+    return -1;
+  }
+
+  HASH_ITER(hh, properties, current, tmp) {
+    if(writeUTF(current->name, strlen(current->name), False, client->_peer->_stream) <= 0) {
+      return -1;
+    }
+    if(writeUTF(current->value, strlen(current->value), False, client->_peer->_stream) <= 0) {
+      return -1;
+    }
+    //client->logger_->log_debug("Site2Site Protocol Send handshake properties %s %s", current->name, current->value);
+    HASH_DEL(properties, current);
```

Comment:
In the cases where return -1 is used, no free is called. Is that going to cause a memory leak?

Reply:
Yes, it would. Fixed, thanks for spotting!